File: test_examples/go/test_stdlib_usage.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"math"
	"math/rand"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"
)

// TestResult represents the result of a test function
type TestResult struct {
	Name    string      `json:"name"`
	Success bool        `json:"success"`
	Data    interface{} `json:"data,omitempty"`
	Error   string      `json:"error,omitempty"`
}

// Test basic Go built-in functions and types
func testBuiltins() TestResult {
	slice := []int{1, 2, 3, 4, 5}
	
	// Test slice operations
	result := map[string]interface{}{
		"len":    len(slice),
		"cap":    cap(slice),
		"append": append(slice, 6),
		"copy": func() []int {
			dest := make([]int, len(slice))
			copy(dest, slice)
			return dest
		}(),
	}
	
	// Test map operations
	m := make(map[string]int)
	m["key1"] = 10
	m["key2"] = 20
	result["map_len"] = len(m)
	result["map_values"] = m
	
	// Test channel operations
	ch := make(chan int, 2)
	ch <- 1
	ch <- 2
	close(ch)
	
	var channelValues []int
	for v := range ch {
		channelValues = append(channelValues, v)
	}
	result["channel_values"] = channelValues
	
	return TestResult{
		Name:    "Built-ins",
		Success: true,
		Data:    result,
	}
}

// Test string operations
func testStringOperations() TestResult {
	text := "Hello, World!"
	
	result := map[string]interface{}{
		"upper":       strings.ToUpper(text),
		"lower":       strings.ToLower(text),
		"replace":     strings.Replace(text, "World", "Go", 1),
		"contains":    strings.Contains(text, "World"),
		"split":       strings.Split(text, ","),
		"join":        strings.Join([]string{"a", "b", "c"}, "-"),
		"trim":        strings.TrimSpace("  spaces  "),
		"has_prefix":  strings.HasPrefix(text, "Hello"),
		"has_suffix":  strings.HasSuffix(text, "!"),
		"repeat":      strings.Repeat("Go", 3),
	}
	
	return TestResult{
		Name:    "String Operations",
		Success: true,
		Data:    result,
	}
}

// Test fmt package
func testFmtPackage() TestResult {
	name := "Go"
	age := 15
	
	result := map[string]interface{}{
		"sprintf":  fmt.Sprintf("Hello, %s! You are %d years old.", name, age),
		"sprint":   fmt.Sprint("Hello", " ", "World"),
		"sprintln": strings.TrimSuffix(fmt.Sprintln("Line", "with", "newline"), "\n"),
	}
	
	return TestResult{
		Name:    "Fmt Package",
		Success: true,
		Data:    result,
	}
}

// Test math package
func testMathPackage() TestResult {
	result := map[string]interface{}{
		"pi":       math.Pi,
		"e":        math.E,
		"sqrt":     math.Sqrt(16),
		"pow":      math.Pow(2, 3),
		"sin":      math.Sin(math.Pi / 2),
		"cos":      math.Cos(0),
		"log":      math.Log(math.E),
		"ceil":     math.Ceil(4.2),
		"floor":    math.Floor(4.8),
		"abs":      math.Abs(-5.5),
		"max":      math.Max(10, 20),
		"min":      math.Min(10, 20),
		"mod":      math.Mod(7, 3),
		"round":    math.Round(4.7),
		"trunc":    math.Trunc(4.7),
	}
	
	return TestResult{
		Name:    "Math Package",
		Success: true,
		Data:    result,
	}
}

// Test time package
func testTimePackage() TestResult {
	now := time.Now()
	
	// Create a specific time for consistent testing
	specificTime := time.Date(2024, 12, 25, 15, 30, 45, 0, time.UTC)
	
	result := map[string]interface{}{
		"now_year":       now.Year(),
		"now_month":      int(now.Month()),
		"now_day":        now.Day(),
		"format":         specificTime.Format("2006-01-02 15:04:05"),
		"unix":           specificTime.Unix(),
		"weekday":        specificTime.Weekday().String(),
		"add_duration":   specificTime.Add(24 * time.Hour).Format("2006-01-02"),
		"since_specific": time.Since(specificTime) > 0,
		"parse_test": func() bool {
			parsed, err := time.Parse("2006-01-02", "2024-12-25")
			return err == nil && parsed.Year() == 2024
		}(),
	}
	
	return TestResult{
		Name:    "Time Package",
		Success: true,
		Data:    result,
	}
}

// Test rand package
func testRandPackage() TestResult {
	rand.Seed(42) // For reproducible results
	
	slice := []string{"apple", "banana", "cherry", "date"}
	rand.Shuffle(len(slice), func(i, j int) {
		slice[i], slice[j] = slice[j], slice[i]
	})
	
	result := map[string]interface{}{
		"intn":       rand.Intn(100),
		"float64":    rand.Float64(),
		"perm":       rand.Perm(5),
		"shuffled":   slice,
		"seed_test":  rand.Intn(1000),
	}
	
	return TestResult{
		Name:    "Rand Package",
		Success: true,
		Data:    result,
	}
}

// Test sort package
func testSortPackage() TestResult {
	ints := []int{3, 1, 4, 1, 5, 9, 2, 6}
	strings := []string{"banana", "apple", "cherry", "date"}
	
	sort.Ints(ints)
	sort.Strings(strings)
	
	// Test custom sort
	people := []struct {
		Name string
		Age  int
	}{
		{"Alice", 30},
		{"Bob", 25},
		{"Charlie", 35},
	}
	
	sort.Slice(people, func(i, j int) bool {
		return people[i].Age < people[j].Age
	})
	
	result := map[string]interface{}{
		"sorted_ints":    ints,
		"sorted_strings": strings,
		"is_sorted":      sort.IntsAreSorted(ints),
		"search_result":  sort.SearchInts(ints, 5),
		"custom_sort":    people,
	}
	
	return TestResult{
		Name:    "Sort Package",
		Success: true,
		Data:    result,
	}
}

// Test strconv package
func testStrconvPackage() TestResult {
	result := map[string]interface{}{}
	
	// String to int
	if i, err := strconv.Atoi("42"); err == nil {
		result["atoi"] = i
	}
	
	// Int to string
	result["itoa"] = strconv.Itoa(42)
	
	// String to float
	if f, err := strconv.ParseFloat("3.14", 64); err == nil {
		result["parse_float"] = f
	}
	
	// Bool operations
	if b, err := strconv.ParseBool("true"); err == nil {
		result["parse_bool"] = b
	}
	result["format_bool"] = strconv.FormatBool(true)
	
	// Base conversion
	result["format_int_base2"] = strconv.FormatInt(42, 2)
	result["format_int_base16"] = strconv.FormatInt(42, 16)
	
	if i, err := strconv.ParseInt("101010", 2, 64); err == nil {
		result["parse_int_base2"] = i
	}
	
	return TestResult{
		Name:    "Strconv Package",
		Success: true,
		Data:    result,
	}
}

// Test os package
func testOsPackage() TestResult {
	result := map[string]interface{}{}
	
	// Get current working directory
	if wd, err := os.Getwd(); err == nil {
		result["getwd"] = wd
	}
	
	// Environment variables
	result["env_home"] = os.Getenv("HOME")
	result["env_path_exists"] = os.Getenv("PATH") != ""
	
	// File operations
	result["temp_dir"] = os.TempDir()
	
	// Args
	result["args_len"] = len(os.Args)
	
	// Hostname
	if hostname, err := os.Hostname(); err == nil {
		result["hostname"] = hostname
	}
	
	return TestResult{
		Name:    "OS Package",
		Success: true,
		Data:    result,
	}
}

// Test filepath package
func testFilepathPackage() TestResult {
	path := "/home/user/documents/file.txt"
	
	dir, file := filepath.Split(path)
	
	result := map[string]interface{}{
		"base":      filepath.Base(path),
		"dir":       filepath.Dir(path),
		"ext":       filepath.Ext(path),
		"join":      filepath.Join("home", "user", "file.txt"),
		"clean":     filepath.Clean("home//user/../user/./file.txt"),
		"abs_test":  filepath.IsAbs(path),
		"split_dir": dir,
		"split_file": file,
		"match": func() bool {
			matched, err := filepath.Match("*.txt", "file.txt")
			return err == nil && matched
		}(),
	}
	
	return TestResult{
		Name:    "Filepath Package",
		Success: true,
		Data:    result,
	}
}

// Test regexp package
func testRegexpPackage() TestResult {
	text := "The year 2024 was great, but 2025 will be better!"
	
	// Simple pattern matching
	matched, _ := regexp.MatchString(`\d{4}`, text)
	
	// Compile and use regexp
	re := regexp.MustCompile(`\d{4}`)
	years := re.FindAllString(text, -1)
	
	// Replace with regexp
	replaced := re.ReplaceAllString(text, "YEAR")
	
	// Split with regexp
	words := regexp.MustCompile(`\s+`).Split(text, -1)
	
	result := map[string]interface{}{
		"match_found":    matched,
		"found_years":    years,
		"replaced_text":  replaced,
		"word_count":     len(words),
		"first_word":     words[0],
		"last_word":      words[len(words)-1],
	}
	
	return TestResult{
		Name:    "Regexp Package",
		Success: true,
		Data:    result,
	}
}

// Test JSON encoding/decoding
func testJSONPackage() TestResult {
	type Person struct {
		Name    string   `json:"name"`
		Age     int      `json:"age"`
		Hobbies []string `json:"hobbies"`
	}
	
	person := Person{
		Name:    "Alice",
		Age:     30,
		Hobbies: []string{"reading", "programming", "hiking"},
	}
	
	// Marshal to JSON
	jsonData, err := json.Marshal(person)
	if err != nil {
		return TestResult{
			Name:    "JSON Package",
			Success: false,
			Error:   err.Error(),
		}
	}
	
	// Unmarshal from JSON
	var decoded Person
	err = json.Unmarshal(jsonData, &decoded)
	if err != nil {
		return TestResult{
			Name:    "JSON Package",
			Success: false,
			Error:   err.Error(),
		}
	}
	
	result := map[string]interface{}{
		"json_string":    string(jsonData),
		"decoded_name":   decoded.Name,
		"decoded_age":    decoded.Age,
		"hobby_count":    len(decoded.Hobbies),
		"round_trip_ok":  person.Name == decoded.Name && person.Age == decoded.Age,
	}
	
	return TestResult{
		Name:    "JSON Package",
		Success: true,
		Data:    result,
	}
}

// Test file I/O operations
func testFileOperations() TestResult {
	testContent := "Hello, Go file I/O!"
	filename := "test_file.txt"
	
	// Write to file
	err := ioutil.WriteFile(filename, []byte(testContent), 0644)
	if err != nil {
		return TestResult{
			Name:    "File Operations",
			Success: false,
			Error:   "Write failed: " + err.Error(),
		}
	}
	
	// Read from file
	content, err := ioutil.ReadFile(filename)
	if err != nil {
		return TestResult{
			Name:    "File Operations",
			Success: false,
			Error:   "Read failed: " + err.Error(),
		}
	}
	
	// Clean up
	os.Remove(filename)
	
	result := map[string]interface{}{
		"write_success":  true,
		"read_content":   string(content),
		"content_match":  string(content) == testContent,
		"file_size":      len(content),
	}
	
	return TestResult{
		Name:    "File Operations",
		Success: true,
		Data:    result,
	}
}

// Test complex operations combining multiple packages
func testComplexOperations() TestResult {
	// Create sample data
	type Product struct {
		Name     string  `json:"name"`
		Price    float64 `json:"price"`
		Category string  `json:"category"`
	}
	
	products := []Product{
		{"Laptop", 999.99, "Electronics"},
		{"Book", 29.99, "Education"},
		{"Coffee", 4.99, "Food"},
		{"Mouse", 19.99, "Electronics"},
		{"Notebook", 5.99, "Education"},
	}
	
	// Sort by price
	sort.Slice(products, func(i, j int) bool {
		return products[i].Price < products[j].Price
	})
	
	// Group by category
	categories := make(map[string][]Product)
	for _, product := range products {
		categories[product.Category] = append(categories[product.Category], product)
	}
	
	// Calculate statistics
	var totalPrice float64
	for _, product := range products {
		totalPrice += product.Price
	}
	avgPrice := totalPrice / float64(len(products))
	
	// Filter expensive items (> $20)
	var expensive []Product
	for _, product := range products {
		if product.Price > 20.0 {
			expensive = append(expensive, product)
		}
	}
	
	// Generate JSON report
	report := map[string]interface{}{
		"total_products":    len(products),
		"average_price":     math.Round(avgPrice*100) / 100,
		"cheapest_product":  products[0].Name,
		"most_expensive":    products[len(products)-1].Name,
		"categories":        len(categories),
		"expensive_count":   len(expensive),
	}
	
	jsonReport, _ := json.Marshal(report)
	
	result := map[string]interface{}{
		"products_sorted":   len(products) == 5,
		"categories_found":  len(categories),
		"avg_price":         avgPrice,
		"json_report":       string(jsonReport),
		"expensive_items":   len(expensive),
	}
	
	return TestResult{
		Name:    "Complex Operations",
		Success: true,
		Data:    result,
	}
}

// Test error handling
func testErrorHandling() TestResult {
	result := map[string]interface{}{}
	
	// Test file not found error
	_, err := ioutil.ReadFile("nonexistent_file.txt")
	result["file_error_handled"] = err != nil
	
	// Test JSON parsing error
	var data interface{}
	err = json.Unmarshal([]byte("invalid json"), &data)
	result["json_error_handled"] = err != nil
	
	// Test strconv parsing error
	_, err = strconv.Atoi("not_a_number")
	result["strconv_error_handled"] = err != nil
	
	// Test regex compilation error
	_, err = regexp.Compile("[invalid")
	result["regex_error_handled"] = err != nil
	
	return TestResult{
		Name:    "Error Handling",
		Success: true,
		Data:    result,
	}
}

func main() {
	fmt.Println("Testing Go Standard Library Usage in Compiled Code")
	fmt.Println(strings.Repeat("=", 60))
	
	tests := []func() TestResult{
		testBuiltins,
		testStringOperations,
		testFmtPackage,
		testMathPackage,
		testTimePackage,
		testRandPackage,
		testSortPackage,
		testStrconvPackage,
		testOsPackage,
		testFilepathPackage,
		testRegexpPackage,
		testJSONPackage,
		testFileOperations,
		testComplexOperations,
		testErrorHandling,
	}
	
	var results []TestResult
	successful := 0
	
	for _, test := range tests {
		result := test()
		results = append(results, result)
		
		fmt.Printf("\nTesting %s...\n", result.Name)
		if result.Success {
			fmt.Printf("✓ %s completed successfully\n", result.Name)
			successful++
		} else {
			fmt.Printf("✗ %s failed: %s\n", result.Name, result.Error)
		}
	}
	
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("Test Summary:")
	fmt.Printf("Total tests: %d\n", len(tests))
	fmt.Printf("Successful: %d\n", successful)
	fmt.Printf("Failed: %d\n", len(tests)-successful)
	
	// Save results to JSON file
	jsonResults, err := json.MarshalIndent(results, "", "  ")
	if err == nil {
		err = ioutil.WriteFile("go_stdlib_test_results.json", jsonResults, 0644)
		if err == nil {
			fmt.Println("\nDetailed results saved to 'go_stdlib_test_results.json'")
		}
	}
	
	if successful == len(tests) {
		fmt.Println("\n🎉 All tests passed!")
		os.Exit(0)
	} else {
		fmt.Printf("\n❌ %d tests failed\n", len(tests)-successful)
		os.Exit(1)
	}
}

File: test_examples/python/test_stdlib_usage.py
#!/usr/bin/env python3
"""
Comprehensive test for Python standard library usage in compiled output.
This tests that the compiler can handle and correctly compile Python code
that uses various standard library modules.
"""

import os
import sys
import json
import math
import datetime
import random
import collections
import itertools
import functools
from typing import List, Dict, Optional, Tuple

# Test basic built-in functions
def test_builtins():
    """Test basic built-in functions"""
    numbers = [1, 2, 3, 4, 5]
    result = {
        'len': len(numbers),
        'sum': sum(numbers),
        'max': max(numbers),
        'min': min(numbers),
        'sorted': sorted(numbers, reverse=True),
        'enumerate': list(enumerate(numbers)),
        'zip': list(zip(numbers, ['a', 'b', 'c', 'd', 'e'])),
        'range': list(range(5)),
        'map': list(map(lambda x: x * 2, numbers)),
        'filter': list(filter(lambda x: x % 2 == 0, numbers))
    }
    return result

# Test string operations
def test_string_operations():
    """Test string manipulation functions"""
    text = "Hello, World!"
    return {
        'upper': text.upper(),
        'lower': text.lower(),
        'replace': text.replace('World', 'Python'),
        'split': text.split(','),
        'join': '-'.join(['a', 'b', 'c']),
        'strip': "  spaces  ".strip(),
        'format': "Hello, {}!".format("Python"),
        'f_string': f"Hello, {'Python'}!"
    }

# Test os module
def test_os_module():
    """Test os module functionality"""
    current_dir = os.getcwd()
    return {
        'getcwd': current_dir,
        'listdir_exists': len(os.listdir('.')) > 0,
        'path_join': os.path.join('path', 'to', 'file.txt'),
        'path_exists': os.path.exists('.'),
        'environ_home': os.environ.get('HOME', 'not_found')
    }

# Test sys module
def test_sys_module():
    """Test sys module functionality"""
    return {
        'platform': sys.platform,
        'version_info': f"{sys.version_info.major}.{sys.version_info.minor}",
        'argv_length': len(sys.argv),
        'maxsize': sys.maxsize > 0
    }

# Test json module
def test_json_module():
    """Test JSON serialization/deserialization"""
    data = {'name': 'test', 'value': 42, 'items': [1, 2, 3]}
    json_str = json.dumps(data)
    parsed = json.loads(json_str)
    return {
        'serialized': json_str,
        'deserialized': parsed,
        'round_trip_success': data == parsed
    }

# Test math module
def test_math_module():
    """Test mathematical functions"""
    return {
        'pi': math.pi,
        'e': math.e,
        'sqrt': math.sqrt(16),
        'pow': math.pow(2, 3),
        'sin': math.sin(math.pi / 2),
        'cos': math.cos(0),
        'log': math.log(math.e),
        'factorial': math.factorial(5),
        'gcd': math.gcd(12, 18),
        'ceil': math.ceil(4.2),
        'floor': math.floor(4.8)
    }

# Test datetime module
def test_datetime_module():
    """Test datetime functionality"""
    now = datetime.datetime.now()
    date_today = datetime.date.today()
    time_now = datetime.time(12, 30, 45)
    
    return {
        'now_year': now.year,
        'today_day': date_today.day,
        'time_hour': time_now.hour,
        'strftime': now.strftime('%Y-%m-%d'),
        'timedelta': str(datetime.timedelta(days=7))
    }

# Test random module
def test_random_module():
    """Test random number generation"""
    random.seed(42)  # For reproducible results
    return {
        'randint': random.randint(1, 100),
        'random': random.random(),
        'choice': random.choice(['a', 'b', 'c', 'd']),
        'shuffle_test': random.sample(range(10), 5)
    }

# Test collections module
def test_collections_module():
    """Test collections data structures"""
    counter = collections.Counter(['a', 'b', 'a', 'c', 'b', 'a'])
    defaultdict_test = collections.defaultdict(int)
    defaultdict_test['key'] += 1
    
    deque_test = collections.deque([1, 2, 3])
    deque_test.appendleft(0)
    deque_test.append(4)
    
    return {
        'counter': dict(counter),
        'counter_most_common': counter.most_common(2),
        'defaultdict': dict(defaultdict_test),
        'deque': list(deque_test)
    }

# Test itertools module
def test_itertools_module():
    """Test itertools functionality"""
    return {
        'chain': list(itertools.chain([1, 2], [3, 4])),
        'cycle_take_5': list(itertools.islice(itertools.cycle(['a', 'b']), 5)),
        'count_take_3': list(itertools.islice(itertools.count(10, 2), 3)),
        'combinations': list(itertools.combinations([1, 2, 3], 2)),
        'permutations': list(itertools.permutations([1, 2, 3], 2)),
        'product': list(itertools.product(['A', 'B'], [1, 2]))
    }

# Test functools module
def test_functools_module():
    """Test functools functionality"""
    @functools.lru_cache(maxsize=128)
    def cached_fibonacci(n):
        if n < 2:
            return n
        return cached_fibonacci(n-1) + cached_fibonacci(n-2)
    
    # Test reduce
    numbers = [1, 2, 3, 4, 5]
    product = functools.reduce(lambda x, y: x * y, numbers)
    
    # Test partial
    multiply_by_2 = functools.partial(lambda x, y: x * y, 2)
    
    return {
        'fibonacci_10': cached_fibonacci(10),
        'reduce_product': product,
        'partial_result': multiply_by_2(5),
        'cache_info': str(cached_fibonacci.cache_info())
    }

# Test complex data structures and algorithms
def test_complex_operations():
    """Test complex operations that combine multiple stdlib features"""
    data = [
        {'name': 'Alice', 'age': 30, 'score': 85},
        {'name': 'Bob', 'age': 25, 'score': 92},
        {'name': 'Charlie', 'age': 35, 'score': 78},
    ]
    
    # Sort by score, descending
    sorted_by_score = sorted(data, key=lambda x: x['score'], reverse=True)
    
    # Group by age range
    age_groups = collections.defaultdict(list)
    for person in data:
        age_range = "young" if person['age'] < 30 else "old"
        age_groups[age_range].append(person['name'])
    
    # Calculate statistics
    scores = [person['score'] for person in data]
    stats = {
        'mean': sum(scores) / len(scores),
        'min': min(scores),
        'max': max(scores),
        'std_dev': math.sqrt(sum((x - sum(scores)/len(scores))**2 for x in scores) / len(scores))
    }
    
    return {
        'sorted_data': sorted_by_score,
        'age_groups': dict(age_groups),
        'statistics': stats,
        'json_export': json.dumps(data, indent=2)
    }

# Test error handling with stdlib
def test_error_handling():
    """Test error handling with standard library operations"""
    results = {}
    
    # Test file operations
    try:
        with open('nonexistent_file.txt', 'r') as f:
            content = f.read()
        results['file_error'] = False
    except FileNotFoundError:
        results['file_error'] = True
    
    # Test JSON parsing error
    try:
        json.loads('invalid json')
        results['json_error'] = False
    except json.JSONDecodeError:
        results['json_error'] = True
    
    # Test math domain error
    try:
        math.sqrt(-1)
        results['math_error'] = False
    except ValueError:
        results['math_error'] = True
    
    return results

def main():
    """Main test function that runs all standard library tests"""
    print("Testing Python Standard Library Usage in Compiled Code")
    print("=" * 60)
    
    test_functions = [
        ("Built-in Functions", test_builtins),
        ("String Operations", test_string_operations),
        ("OS Module", test_os_module),
        ("Sys Module", test_sys_module),
        ("JSON Module", test_json_module),
        ("Math Module", test_math_module),
        ("DateTime Module", test_datetime_module),
        ("Random Module", test_random_module),
        ("Collections Module", test_collections_module),
        ("Itertools Module", test_itertools_module),
        ("Functools Module", test_functools_module),
        ("Complex Operations", test_complex_operations),
        ("Error Handling", test_error_handling)
    ]
    
    all_results = {}
    
    for test_name, test_func in test_functions:
        print(f"\nTesting {test_name}...")
        try:
            result = test_func()
            all_results[test_name] = result
            print(f"✓ {test_name} completed successfully")
        except Exception as e:
            print(f"✗ {test_name} failed: {e}")
            all_results[test_name] = {"error": str(e)}
    
    print("\n" + "=" * 60)
    print("Test Summary:")
    print(f"Total tests: {len(test_functions)}")
    
    successful_tests = sum(1 for result in all_results.values() if "error" not in result)
    print(f"Successful: {successful_tests}")
    print(f"Failed: {len(test_functions) - successful_tests}")
    
    # Save results to JSON file
    with open('python_stdlib_test_results.json', 'w') as f:
        json.dump(all_results, f, indent=2, default=str)
    
    print("\nDetailed results saved to 'python_stdlib_test_results.json'")
    
    return successful_tests == len(test_functions)

if __name__ == "__main__":
    success = main()
    sys.exit(0 if success else 1)